src/decoder/image.rs

```diff
@@ -447,6 +449,25 @@ impl Image {
 
                 Box::new(Cursor::new(data))
             }
+            CompressionMethod::Fax4 => {
+                let width = u16::try_from(self.width)?;
+                let height = u16::try_from(self.height)?;
+                let mut out: Vec<u8> = Vec::with_capacity(usize::from(width) * usize::from(height));
```

Comment:
I have two concerns here:
1. This is likely to cause fuzzer OOMs because just setting a few header fields in enough to cause a ~4GB allocation.
2. The size calculation assumes that each pixel takes one byte, but I'd expect `Gray(1)` to mean one _bit_ per pixel

Reply:
Hmm. Going to push back a bit on (2), but you're the library maintainer so LMK if I'm looking at this wrong. I expect that this crate reports back what the tiff file says in its header, which *is* `Gray(1)`. Then in the outer crate (`image`) we'd do this:

```
diff --git a/src/codecs/tiff.rs b/src/codecs/tiff.rs
index 9f4dd735..256215c5 100644
--- a/src/codecs/tiff.rs
+++ b/src/codecs/tiff.rs
@@ -53,6 +53,7 @@ where
         };

         let color_type = match tiff_color_type {
+            tiff::ColorType::Gray(1) => ColorType::L8,
```

Does that make sense? The alternatives would be to alter the following to special case this and override what's coming from the header:  https://github.com/image-rs/image-tiff/blob/master/src/decoder/image.rs#L351

Alternatively we could add an extra data type to represent the actual output.

I'm OK with any of these options, just want to lay them out. What would you like as both the author/maintainer of this library *and* the primary consumer of its API in the `image` crate?

Reply:
This crate should already decode `Gray(1)` encoded images into packed 1-bit per sample outputs. 

To support the image crate use case, I'd expect that this create should either exposes an _optional_ flag to expand sub 8-bit channels ([like PNG does](https://docs.rs/png/latest/png/struct.Transformations.html#associatedconstant.EXPAND)), or else not perform any expansion and have code in the image crate to do the conversion from packed 1-bit per sample into L8 encoded.

Reply:
I'm working at getting packed 1-bit samples working here and, TBH, it's quite a bit of fiddly work. I've got the decoder portion outputting 1-bit samples but the rest of the crate seems to assume it's byte-addressable.  I am looking at the chunking (de-chunking?) code in `fn expand_chunk` and there's a lot that assumes samples are individually addressable. Should I go and start altering this to change this assumption? Is this something you could provide more guidance on?

To step back a bit: especially if the `image` crate, presumably the main consumer of this, just goes and re-expands the samples back to individual bytes, how much does it offer to have this crate offer 1-bit samples? There's a pretty straightforward path to getting this working in the `image` crate with this decoder without worrying about any of this.

Reply:
OK, I've looked a bit further and `expand_chunk` is the difficult part here.

There are three different code paths

Reply:
I realized the reason I thought there was sub-byte sample support was because there's a [stalled-out PR](#218) for it that I lost track of.

I'll hopefully have a bit more time this weekend, but briefly:

* My recollection is that TIFF packs multiple samples per byte across pixels but each row is always an integer number of bytes (i.e. it is padded out to a multiple of 8 bits). 
* To handle the first case of the chunk width matching the image width, the byte layout should match what's in the file so you can just copy byte-by-byte. 
* The floating point predictor implies floating point samples, which are always 16-bit, 32-bit, or 64-bit. Thus you don't have to handle it.
* The final case can only be hit for tiled images. For here, I think it would be reasonable to return an unsupported error if the tile_width * bit_depth isn't a multiple of 8. With that assumption I don't think the code should be too hard

A final note is to watch out for integer overflow when using `u32` or `usize`. It only takes 512 MB to store 2^32 bits, so a 32-bit system could plausibly want to decode a TIFF file containing more pixels than fit in a `usize`. And even if the decoded image won't fit in RAM, we still want to be able to return a graceful error rather than panicking due to integer overflow

Reply:
OK. I'll be away all next week but I'd like to pick this up again after that. Thanks for the context here.